WIP: adding unit tests

DucAnh-silabs opened the pull request:
Relate-to: https://github.com/SiliconLabsSoftware/z-wave-nvm-migration-tool/issues/2

rzr commented on tests/controllerlib/zwave/test_ZW_nodemask.c:
```diff
@@ -0,0 +1,96 @@
+#include "unity.h"
```
it should be system path <unity.h>

DucAnh-silabs replied:
I pull the cmock source and build it, should I still set it to system path <unity.h> ?

rzr replied:
yes it is external to project (in a library)

rzr commented:
I see that 

✖ Quality Gate has FAILED.

Detailed information can be found at: s.s.n/dashboard?id=z-wave-nvm-migration-tool&pullRequest=9

Are you able to troubleshoot it ?

DucAnh-silabs commented:
Hi @rzr, It was the 0% coverage that marked it as failed

rzr commented:
> Hi @rzr, It was the 0% coverage that marked it as failed

ok let me try to fix that with change i made elsewhere

you can refer to this project it does the scanning properly, you can try locally:

https://github.com/SiliconLabsSoftware/z-wave-protocol-controller/pull/134

Try:

    $ ./helper.mk SONAR_OUT_DIR=sonar sonar/setup default coverage

Let me share the same helper.mk it can help to avoid relying on github

btw when you are copying code other place it is good to add reference to the source in the commit message that way we can align easier if needed.

DucAnh-silabs committed:
Adding unit tests
build.yml: add ruby package to generate mock with cmock

mod(sonar): exclude tests/ and nvm3lib/ from source analysis

rzr commented:
I checked gate is passing on PRs

https://github.com/SiliconLabsSoftware/z-wave-nvm-migration-tool/actions/runs/16318733626

Please approve:

https://github.com/SiliconLabsSoftware/z-wave-nvm-migration-tool/pull/10

I would suggest you to isolate your changes in smaller chunks and get the CI passing

Eventually just add a simple ctest and see if it covered (without sonar).

./helper.mk COVERAGE=ON all test coverage 

should report something just like in: 

https://github.com/SiliconLabsSoftware/z-wave-protocol-controller/pull/134 

Ask me hints if unclear

rzr commented on sonar-project.properties:
```diff
@@ -4,3 +4,5 @@ sonar.projectKey=z-wave-nvm-migration-tool
 
 sonar.exclusions=build/
+sonar.exclusions=tests/**/*
+sonar.exclusions=zw_nvm_converter/nvm3lib/**/*
```
does this work?

I thought it should be on same line with , separators

BTW: 

```
tests/**/* ->  test*/**
```

rzr commented:
Please rebase., if there is anything blocked please let me know I will try to help

DucAnh-silabs commented on sonar-project.properties:
```diff
@@ -3,4 +3,4 @@ sonar.projectKey=z-wave-nvm-migration-tool
 # --- optional properties ---
 
-sonar.exclusions=build/
+sonar.exclusions=build/, test*/**/*, zw_nvm_converter/nvm3lib/**/*
```
Hi @rzr , 
Is this format okay? I saw that the code coverage is not computed for `tests` folder in the report. BTW, how do I compute the coverage on source in order for it to display on sonar dashboard?

rzr replied:
```
sonar.exclusions=build/,test*/**, zw_nvm_converter/nvm3lib/**
```

DucAnh-silabs replied:
Okay, I'll modify it

rzr commented on .github/workflows/build.yml:
```diff
@@ -26,7 +26,7 @@ jobs:
           sudo apt-get install -y ${{ env.debian_packages }}
         env:
-          debian_packages: cmake build-essential gcovr libjson-c-dev
+          debian_packages: cmake build-essential gcovr libjson-c-dev ruby 
```
please add it in helper.mk too

or replace above lines by 

helper.mk setup

rzr commented on tests/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+                     ${cmock_SOURCE_DIR}/vendor/unity/src
+)
+# set(COMMON_TEST_SRC 
+#   ${CMAKE_SOURCE_DIR}/tests
```
please remove

rzr commented on tests/controllerlib/test_api_wrappers.c:
Please add a comment  why it is needed

DucAnh-silabs replied:
This one is a test file for `api_wrappers.c`, I add it to the commit by mistake.

rzr reviewed (commented):
I assume you plan to do some cleanup. but it looks it is passing now good job

DucAnh-silabs commented:
There are still more tests I have to add, and some clean up too.

rzr commented:
> There are still more tests I have to add, and some clean up too.

ok but you can add them later, if it is passing that is good enough to merge then we can rebase your other PR and see if any blocker for release

DucAnh-silabs commented:
Okay

rzr commented on helper.mk:
```diff
@@ -39,6 +39,7 @@ debian_packages+=build-essential
 debian_packages+=cmake
 # packages/deps/libs:
 debian_packages+=libjson-c-dev
+debian_packages+=ruby
```
i would put it in later section since it is not a lib to build the app

DucAnh-silabs committed:
Add ruby package to helper.mk and using helper.mk for github action

rzr commented on sonar-project.properties:
```diff
@@ -3,4 +3,4 @@ sonar.projectKey=z-wave-nvm-migration-tool
 # --- optional properties ---
 
-sonar.exclusions=build/
+sonar.exclusions=build/, test*/**, zw_nvm_converter/nvm3lib/**
```
```suggestion
sonar.exclusions=build/,test*/**,zw_nvm_converter/nvm3lib/**
```

rzr commented on tests/controllerlib/CMakeLists.txt:
```diff
@@ -0,0 +1,74 @@
+    ${CMAKE_SOURCE_DIR}/zw_nvm_converter/nvm3lib/platform/emdrv/common/inc/
+		${CMAKE_SOURCE_DIR}/zw_nvm_converter/controllerlib_7.xx/zwave
+)
+target_compile_definitions(nvm3_mock PUBLIC NVM3_HOST_BUILD FLASH_PAGE_SIZE=2048 NVM3_DEFAULT_CACHE_SIZE=400)
```
extra comment would help to explain why it needs to be overloaded , if not needed please remove

rzr commented on tests/controllerlib/CMakeLists.txt:
```diff
@@ -0,0 +1,74 @@
+    ${CMAKE_SOURCE_DIR}/zw_nvm_converter/nvm3lib/platform/emdrv/nvm3/inc/
+    ${CMAKE_SOURCE_DIR}/zw_nvm_converter/nvm3lib/platform/emdrv/nvm3/config/
+    ${CMAKE_SOURCE_DIR}/zw_nvm_converter/nvm3lib/platform/emdrv/common/inc/
+		${CMAKE_SOURCE_DIR}/zw_nvm_converter/controllerlib_7.xx/zwave
```
weird indent ? Are you using tabs ? please prefer spaces (double?)

rzr commented on tests/controllerlib/CMakeLists.txt:
```diff
@@ -0,0 +1,74 @@
+target_link_libraries(test_nvm3_helpers 
+                      unity
+                      nvm3_mock
+											user_message_mock
```
nitpick: style is odd

rzr commented on tests/controllerlib/CMakeLists.txt:
```diff
@@ -0,0 +1,74 @@
+add_test(NAME test_nvm3_helpers COMMAND test_nvm3_helpers)
+
+add_executable(test_controller_nvm test_controller_nvm.c
+            # ${CMAKE_SOURCE_DIR}/zw_nvm_converter/controllerlib_7.xx/controller_nvm.c
```
```suggestion
```

rzr commented on tests/controllerlib/test_controller_nvm.c:
```diff
@@ -0,0 +1,829 @@
+  TEST_ASSERT_FALSE(app_nvm_is_pre_v7_21(7, 21, 0));
+}
+
+// void test_check_controller_nvm(void)
```
please add in later PR (As draft eventually)

rzr reviewed (commented):
I am in favor of reducing the change to merge asap to check sonar status

DucAnh-silabs commented:
Hi @rzr
I've made changes as suggested.

DucAnh-silabs committed:
Reduced changes, leave todos for next PR

rzr reviewed (approved):
Please mark it ready, is it ok to squash the commits ?

DucAnh-silabs commented:
Yes, you can squash it
